api/v1alpha1/rule_types.go

```diff
@@ -47,41 +61,137 @@ type RuleSpec struct {
+	// Methods represent an array of HTTP methods (e.g. GET, POST, PUT, DELETE, ...)
 	Methods []string `json:"methods"`
 }
 
+// Authenticator represents a handler that authenticates provided credentials.
 type Authenticator struct {
 	*Handler `json:",inline"`
 }
 
+// Authorizer represents a handler that authorizes the subject ("user") from the previously validated credentials making the request.
 type Authorizer struct {
 	*Handler `json:",inline"`
 }
 
+// Mutator represents a handler that transforms the HTTP request before forwarding it.
 type Mutator struct {
 	*Handler `json:",inline"`
 }
 
+// Handler represents an Oathkeeper routine that operates on incoming requests. It is used to either validate a request (Authenticator, Authorizer) or modify it (Mutator).
 type Handler struct {
+	// Name is the name of a handler
 	Name string `json:"handler"`
+	// Config configures the handler. Configuration keys vary per handler.
 	// +kubebuilder:validation:Type=object
 	Config *runtime.RawExtension `json:"config,omitempty"`
 }
 
+// ToOathkeeperRules transforms a RuleList object into a JSON object digestible by Oathkeeper.
+func (rl RuleList) ToOathkeeperRules() ([]byte, error) {
+
+	rules := make([]*RuleJSON, len(rl.Items))
+
+	for i, _ := range rl.Items {
+		rules[i] = rl.Items[i].ToRuleJSON()
+	}
+
+	return json.MarshalIndent(rules, "", "  ")
+}
+
+// FilterNotValid filters out Rules which doesn't pass validation due to being not processed yet or due to negative result of validation. It returns a list of Rules which passed validation successfully.
+func (rl RuleList) FilterNotValid() RuleList {
+	rlCopy := rl
+	validRules := []Rule{}
+	for _, rule := range rl.Items {
+		if rule.Status.Validation != nil && rule.Status.Validation.Valid != nil && *rule.Status.Validation.Valid {
+			validRules = append(validRules, rule)
+		}
+	}
+	rlCopy.Items = validRules
+	return rlCopy
+}
+
+// ValidateWith uses provided validation configuration to check whether the rule have proper handlers set.
+func (r Rule) ValidateWith(config validation.Config) error {
+	// if a handler is not provided in any of the following fields, a default one will be used, so there is no need to check if it is valid
+	if r.Spec.Authenticators != nil {
+		for _, authenticator := range r.Spec.Authenticators {
+			if valid := config.IsAuthenticatorValid(authenticator.Name); !valid {
+				return fmt.Errorf("authenticator: %s is invalid", authenticator.Name)
+			}
+		}
+	}
+	if r.Spec.Authorizer != nil {
+		if valid := config.IsAuthorizerValid(r.Spec.Authorizer.Name); !valid {
+			return fmt.Errorf("authorizer: %s is invalid", r.Spec.Authorizer.Name)
+		}
+	}
+	if r.Spec.Mutator != nil {
+		if valid := config.IsMutatorValid(r.Spec.Mutator.Name); !valid {
+			return fmt.Errorf("mutator: %s is invalid", r.Spec.Mutator.Name)
+		}
+	}
+	return nil
+}
+
+// ToRuleJSON transforms a Rule object into an intermediary RuleJSON object
+func (r Rule) ToRuleJSON() *RuleJSON {
+
+	ruleJSON := &RuleJSON{
```

Comment:
Defaults should *never* fall back to something that allows everything. One little misconfiguration (yaml indentation anyone?) can lead to ALL traffic being allowed. Instead, if you want to have defaults, either use `unauthorized` and `deny` and `noop` for authn, authz, mutator - or leave them empty which will cause oathkeeper to throw a 50x erorr message when accessing that rule.

Reply:
Thanks for the tip. Default handlers for Authenticator, Authorizer and Mutator set to "unauthorized", "deny" and "noop", respectively.